### PR TITLE
Custom Decision Categories

### PR DESCRIPTION
**Pull request by Jeff-Thompson12**
Addresses #276

### CODE REVIEW

**Comment by Jeff-Thompson12** [comment]
Alright. I moved the database specifications and decision categories to a configuration file `db-config.yml`.
image

This utilizes the built in `{golem}` configuration options. I have moved the database specifications out of the arguments for `run_app()`. The tests are currently failing because the database needs to be updated. I would like feedback before I do that.

**Review by AARON-CLARK — CHANGES_REQUESTED** [review]
Overall, this PR is 🔥 and the progress should be seriously celebrated. I do have some comments:

First, using a config file is the right thing to do, but it is also "scary-er" to some users steepens the barrier to entry FOR SOME. It's just another new thing they'll have to learn, so anything we can do to make it easier for them would be a win. As such, I think this PR necessitates augmenting the deployment documentation that is "in the works" via PR #450. After this PR is integrated, I'd like you to contribute a whole section dedicated to the config file not only sharing how to use it, but also background info on how config files work.

So, I know that this PR is still in draft mode, so I know it's still a work in progress, but when I started playing with this branch, I noticed a few things you may or may not have gotten around to yet:

(1) When I run as is, with the defaults. I noticed the decision rules for each category are automatically evenly distributed over the 0 - 1 spectrum - so in this case, split into thirds. And if I added one more category, I noticed it broke into fourths. Love that automation - makes the user's lives easier for sure... However, I did notice that the third slider was all red. Not sure why...

image

image



(2) I also saw how easy it is to set up preset decision rules in the example config. I did notice something else we may want to address in this PR: when I tried to set I set my own decision rules for the default config...

image

And then I later decide to add decision rules for the two categories I left out, I noticed they default to their fourths. Is there anyway we could make those categories disburse evenly over the range slider that wasn't accounted for in the config file? If I wasn't paying attention, I would have missed .2 to .25 and .75 to .8.

image





(3) Also, I tried to test with only two categories, and it didn't work either. See error message below. It seems the "minus 2" when selecting a color palette is hurting us. I could see an org wanting to only have a "keep" or "discard" categories (or "compliant" / "not compliant" or Go/No Go, etc).. Honestly, I could see a company just wanting to use one category.

image

image

(4) I know this PR is still draft, but there doesn't seem to be any protection when specifying strange presets for decision rules. For example:

image

OR

image


Both results in borked range sliders. It looks like the ranges are still stored appropriately in the DT table though:

image

And when you try to load a pkg when their is overlap in ranges however, you get this error:

image

In addition, it doesn't handle categories that share the same name well. This action would have to be stopped:

image

image


So, in summary, we may need a way to provide helpful informative error messaging to the console. Plus the documentation would help.


(5) Along the lines of naming the categories, consider this example. I don't think we should orient the app in such a way that assume the suffix of "risk" makes sense. Yes, the decision is "risk-based", but the decision could reflect what the user hopes to do with that info. With the following example:

image

We should remove the word risk here. If the user wants to display the word "risk" they should include it into the config file. It looks like the config file handles decision categories that are multiple words quite well.
image

However, I noticed that we'll need to change the label on this slider (below) to be "Select Package Decision". Also, we might need to add some padding on the left and right as the text was going off screen.

image

(6) thinking more about the example directly above, perhaps we need to give the user an opportunity to change the default colors around manually. For example, if they feel the "needs review" category/tag isn't really on the same risk gradient/spectrum, perhaps they'd want to change the color to grey or something. Would it be possible to insert a collapsible color picker into the Admin Tools tab? I'd think you'd want your ADMIN users to have access to this after the app is deployed so perhaps it doesn't belong in the config file? Or perhaps you can set defaults in the config, but the admin can change those values, and it could get stored in the db?

Phew, that's enough for now. In summary - this PR is top notch and I love where it's going. Nice work!

**Comment by AARON-CLARK** [comment]
Just a loose thought - I did not test what happens when the user changes the auto-decision thresholds during a session. Do the new thresholds get overwritten in the config file? or does it get stored in the `database.sqlite` file and those will get used over the config file, if they exist?

**Comment by Jeff-Thompson12** [comment]
> Just a loose thought - I did not test what happens when the user changes the auto-decision thresholds during a session. Do the new thresholds get overwritten in the config file? or does it get stored in the `database.sqlite` file and those will get used over the config file, if they exist?

@AARON-CLARK It does not change the configuration file. My thoughts were that the configuration file was mostly meant to setup the application on initialization. This also allows it to be shared across projects. You can easily use the same credential and/or decision rules for multiple initializations.

This is also why I added the decision rules to the database itself. Any changes to the rules after initialization should be tracked with that project and shouldn't change configurations for other projects.

**Comment by AARON-CLARK** [comment]
Another thought I had while writing the 'deployments' vignette today:

https://github.com/pharmaR/riskassessment/blob/433ecc9a76a1d5ef582b6f6542b83ef589dbb478/vignettes/Deployment.Rmd#L65

If you read that paragraph, you'll recall that we currently initialize weights using [initialize_metric_table.sql](https://github.com/pharmaR/riskassessment/blob/master/inst/app/www/sql_queries/initialize_metric_table.sql) file to update the `metric` table. Perhaps we should be initializing those in the config file. Thoughts?

**Comment by Jeff-Thompson12** [comment]
> Perhaps we should be initializing those in the config file. Thoughts?

I think that's a good direction to head. Moving some of these items to configurations with defaults.

File: R/mod_decision_automation.R [hunk 1]
```diff
@@ -17,15 +82,16 @@ mod_decision_automation_ui <- function(id){
...
-#' @importFrom jsonlite read_json write_json
 #' @importFrom purrr compact
 #' @importFrom shinyWidgets tooltipOptions
 mod_decision_automation_server <- function(id, user){
   moduleServer( id, function(input, output, session){
     ns <- session$ns
     
-    auto_json <- jsonlite::read_json("auto_decisions.json")
-    auto_list <- reactiveVal(auto_json)
+    auto_db <- process_dec_tbl()
```
[the remark this answers is not in the document]
Comment (Jeff-Thompson12) [reply R/mod_decision_automation.R]: I have done some minor variable renaming. Take a look and let me know what you think.

File: R/run_app.R [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #'   Please make sure name ends with '.sqlite'. For example: 'cred_db.sqlite'.
 #' @param assessment_db_name text string that names the credentials databse.
 #'   Please make sure name ends with '.sqlite'. For example: 'assess_db.sqlite'.
+#' @param decision_categories a character vector of the decision categories
```
[the remark this answers is not in the document]
Comment (Jeff-Thompson12) [reply R/run_app.R]: The unneeded parameters have been removed.

**Comment by Jeff-Thompson12** [comment]
Reorganizing the changes requested for better tracking...

- [ ] Rogue color in slider
> I edited the `sliders.css` file. If you restart the R session the rogue color will go away.

- [ ]  Fill remaining gaps between ranges not covered in config
> It is hard for me to envision how to implement something like this without a bunch of obtuse code that potentially slows down the initialization of the app. I don't really think this is necessary either since we do ask for verification of selected inputs. Of course if we think this would be a feature a user would be looking for, I can still work to implement it.

- [ ] Accommodate two categories or even just one
> This has been added now.

- [ ] We need config file error handling
> Some rudimentary checks have been implemented. Let me know if any situations seem to be missing.

- [ ] Re-naming
> I have removed the automatic "Risk" suffix added to the decision. This will allow initialization of non-Likert style decision categories. The decision automation may not make sense in those situations, but it need not be enabled either.

- [ ] Custom colors
> I agree that we should allow the user to assign their own colors (especially since we have a decision categories database table now), but I think that should be handled in a separate issue/PR.

**Comment by AARON-CLARK** [comment]
> * [x]  Accommodate two categories or even just one
> 
> > This has been added now.

I just tested one category and this doesn't seem to be working yet. Optimally, we'd just swap out the `sliderInput()` for a `shinyWidgets::materialswitch()`

image

**Comment by AARON-CLARK** [comment]
Whoops!

**Comment by AARON-CLARK** [comment]
> * [ ]   Fill remaining gaps between ranges not covered in config
> 
> > It is hard for me to envision how to implement something like this without a bunch of obtuse code that potentially slows down the initialization of the app. I don't really think this is necessary either since we do ask for verification of selected inputs. Of course if we think this would be a feature a user would be looking for, I can still work to implement it.

In this example, I create some presets of [0,.1] and [.5,1]. When clicking to add "Medium", the range that appears just seems really odd. Haha, I think people will be really confused. But if you think my suggestion will negatively impact the app, then I'm willing to live with it. No biggie.

image

**Review by AARON-CLARK — CHANGES_REQUESTED** [review]
Something seems to be broken. I restarted my session, deleted the `database.sqlite` file, and re-ran with this config file:

image

But the sliders came out like below. Can you replicate it?

image

The table came out just fine:

image

**Comment by Jeff-Thompson12** [comment]
@AARON-CLARK The code I previously had should have worked. There is something going on in `{formattable}` when it parses the formatting functions. Making the color list a named function fixes the issue.

**Review by AARON-CLARK — APPROVED** [review]
💪🏼

**Pull request merged**